{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Differential rhytmicity analysis of omics datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "The aim of this tutorial is to use the cosinor_lite Python package to perform differential rhythmicity analysis on omics data. Here we will use published RNA-seq data, but in theory any types of omics data (RNA-seq, proteomics, metabolomics, lipidomics) could be used. The dataset is from the following publication:\n",
    "\n",
    "Petrenko V, Saini C, Giovannoni L, Gobet C, Sage D, Unser M, Heddad Masson M, Gu G, Bosco D, Gachon F, Philippe J, Dibner C. 2017. Pancreatic α- and β-cellular clocks have distinct molecular properties and impact on islet hormone secretion and gene expression. Genes Dev 31:383–398. doi:10.1101/gad.290379.116\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2",
   "metadata": {},
   "source": [
    "### How does the method actually work?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3",
   "metadata": {},
   "source": [
    "The details of the method are nicely explained in the article: \n",
    "\n",
    "Pelikan A, Herzel H, Kramer A, Ananthasubramaniam B. 2022. Venn diagram analysis overestimates the extent of circadian rhythm reprogramming. The FEBS Journal 289:6605–6621. doi:10.1111/febs.16095\n",
    "\n",
    "Here is an adaptation of their figure explaining the methdology:\n",
    "\n",
    "<div style=\"text-align: center;\">\n",
    "  <img src=\"../images/model_selection.png\" alt=\"Model selection\" style=\"width:60%;\"/>\n",
    "</div>\n",
    "\n",
    "For condition 1 (i.e. alpha cells) and condition 2 (i.e. beta cells), we fit 5 different models:\n",
    "\n",
    "- Model 1) Arrhythmic in alpha and beta cells\n",
    "\n",
    "- Model 2) Rhythmic in beta cells only\n",
    "\n",
    "- Model 3) Rhythmic in alpha cells only\n",
    "\n",
    "- Model 4) Rhythmic in alpha and beta cells with the same rhythmic parameters (i.e. phase and amplitude)\n",
    "\n",
    "- Model 5) Rhythmic in both but with differential rhythmicity in alpha vs beta cells\n",
    "\n",
    "A degree of confidence is calculated for each model (called model weight, which sums to 1 across all models), and a model is chosen if the model weight exceeds a threshold (for this tutorial we will use 0.5). If no model exceeds this threshold, then the model is unclassified, which we define as Model 0\n",
    "\n",
    "- Model 0) unclassified\n",
    "\n",
    "Let's start running the code to make things more concrete! First we need to load the required packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from cosinor_lite.omics_dataset import OmicsDataset\n",
    "from cosinor_lite.omics_differential_rhytmicity import DifferentialRhythmicity, OmicsHeatmap, TimeSeriesExample\n",
    "\n",
    "plt.rcParams.update(\n",
    "    {\n",
    "        \"font.size\": 8,\n",
    "        \"axes.titlesize\": 8,\n",
    "        \"axes.labelsize\": 8,\n",
    "        \"xtick.labelsize\": 8,\n",
    "        \"ytick.labelsize\": 8,\n",
    "        \"legend.fontsize\": 8,\n",
    "        \"figure.titlesize\": 8,\n",
    "        \"pdf.fonttype\": 42,\n",
    "    },\n",
    ")\n",
    "plt.style.use(\"seaborn-v0_8-ticks\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "Next, we load in the omics dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "file: Path = Path.cwd().parent / \"data\" / \"GSE95156_Alpha_Beta.txt\"\n",
    "\n",
    "df_rna: pd.DataFrame = pd.read_csv(file, sep=\"\\t\")\n",
    "df_rna: pd.DataFrame = df_rna.drop(df_rna.columns[0], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7",
   "metadata": {},
   "source": [
    "## Defining sample conditions and times\n",
    "\n",
    "Typically in omics datasets, each row represents a different entity (gene/protein/lipid etc.), and each samples are stored as separate column.\n",
    "\n",
    "We need to tell the algorithm the names of the samples corresponding to condition 1 and 2, as well as the corresponding times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_rna[\"Genes\"] = df_rna[\"gene_name\"].str.split(\"|\").str[1]\n",
    "\n",
    "columns_cond1: list[str] = [\n",
    "    \"ZT_0_a_1\",\n",
    "    \"ZT_4_a_1\",\n",
    "    \"ZT_8_a_1\",\n",
    "    \"ZT_12_a_1\",\n",
    "    \"ZT_16_a_1\",\n",
    "    \"ZT_20_a_1\",\n",
    "    \"ZT_0_a_2\",\n",
    "    \"ZT_4_a_2\",\n",
    "    \"ZT_8_a_2\",\n",
    "    \"ZT_12_a_2\",\n",
    "    \"ZT_16_a_2\",\n",
    "    \"ZT_20_a_2\",\n",
    "]\n",
    "\n",
    "columns_cond2: list[str] = [\n",
    "    \"ZT_0_b_1\",\n",
    "    \"ZT_4_b_1\",\n",
    "    \"ZT_8_b_1\",\n",
    "    \"ZT_12_b_1\",\n",
    "    \"ZT_16_b_1\",\n",
    "    \"ZT_20_b_1\",\n",
    "    \"ZT_0_b_2\",\n",
    "    \"ZT_4_b_2\",\n",
    "    \"ZT_8_b_2\",\n",
    "    \"ZT_12_b_2\",\n",
    "    \"ZT_16_b_2\",\n",
    "    \"ZT_20_b_2\",\n",
    "]\n",
    "\n",
    "t_cond1: np.ndarray = np.array(\n",
    "    [\n",
    "        0,\n",
    "        4,\n",
    "        8,\n",
    "        12,\n",
    "        16,\n",
    "        20,\n",
    "        0,\n",
    "        4,\n",
    "        8,\n",
    "        12,\n",
    "        16,\n",
    "        20,\n",
    "    ],\n",
    ")\n",
    "\n",
    "t_cond2: np.ndarray = t_cond1.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9",
   "metadata": {},
   "source": [
    "## Processing the data and visualising the distribution of expression\n",
    "\n",
    "We next load the data and create a plot of the distribution of all expression values.\n",
    "\n",
    "In some cases there will be a bimodal distribution, with one large peak at high values and a second, typically smaller peak at very low values. This can be a sign that the small peak at low values represents noise and can be considered as unexpressed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_data: OmicsDataset = OmicsDataset(\n",
    "    df=df_rna,\n",
    "    columns_cond1=columns_cond1,\n",
    "    columns_cond2=columns_cond2,\n",
    "    t_cond1=t_cond1,\n",
    "    t_cond2=t_cond2,\n",
    "    deduplicate_on_init=True,\n",
    "    log2_transform=False,\n",
    ")\n",
    "fig = rna_data.expression_histogram()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11",
   "metadata": {},
   "source": [
    "## Quality control\n",
    "\n",
    "Another basic technique for quality control is to create a scatterplot of technical replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = rna_data.replicate_scatterplot(sample1=\"ZT_0_a_1\", sample2=\"ZT_0_a_2\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13",
   "metadata": {},
   "source": [
    "In some cases this may reveal that measurements below a certain threshold look visually less correlated between samples. Here we will defines genes as non-expressed using a mean expression cut-off of 0 and requiring at least two detected measurements per condition, matching the original 2017 article."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_data.add_is_expressed(mean_min=0, num_detected_min=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15",
   "metadata": {},
   "source": [
    "## Fitting rhythmic models to the omics dataset\n",
    "\n",
    "Now things get exciting! We can finally start fitting the rhythmic models to the data. The output of calling DifferentialRhythmicity().extract_all_circadian_params() is a dataframe that contains the model weights across the 5 different models, as well as the amplitude and phase from the selected model for each condition. This can take some time (~5 mins), so go and get a coffee!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "dr = DifferentialRhythmicity(dataset=rna_data)\n",
    "rhythmic_all = dr.extract_all_circadian_params()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17",
   "metadata": {},
   "source": [
    "## Heatmap of model selection\n",
    "\n",
    "We can display the results as a heatmap, where we'll visualise models 2-5 using z-scores for each sample.\n",
    "\n",
    "Note that genes expressed that are only expressed in one condition are given the subclass \"a\", while genes that are expressed in both conditions are given the subclass \"b\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmap: OmicsHeatmap = OmicsHeatmap(\n",
    "    df=rhythmic_all,\n",
    "    columns_cond1=columns_cond1,\n",
    "    columns_cond2=columns_cond2,\n",
    "    t_cond1=t_cond1,\n",
    "    t_cond2=t_cond2,\n",
    "    cond1_label=\"Alpha cells\",\n",
    "    cond2_label=\"Beta cells\",\n",
    "    show_unexpressed=False,\n",
    ")\n",
    "fig = heatmap.plot_heatmap()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19",
   "metadata": {},
   "source": [
    "## Showing individual examples\n",
    "\n",
    "Finally, we can manually select genes of interest to show the underlying data (as points) and the fitted cosinor models from the selected model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "example = TimeSeriesExample(\n",
    "    df=rhythmic_all,\n",
    "    columns_cond1=columns_cond1,\n",
    "    columns_cond2=columns_cond2,\n",
    "    t_cond1=t_cond1,\n",
    "    t_cond2=t_cond2,\n",
    "    cond1_label=\"Alpha cells\",\n",
    "    cond2_label=\"Beta cells\",\n",
    ")\n",
    "\n",
    "example.plot_time_series(\"Arntl\", xticks=np.arange(0, 25, 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cosinor-lite",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
